pkg/tablecodec/tablecodec.go

```diff
@@ -1225,14 +1244,35 @@ func GenIndexKey(loc *time.Location, tblInfo *model.TableInfo, idxInfo *model.In
 		return nil, false, err
 	}
 	if !distinct && h != nil {
-		if h.IsInt() {
-			// We choose the efficient path here instead of calling `codec.EncodeKey`
-			// because the int handle must be an int64, and it must be comparable.
-			// This remains correct until codec.encodeSignedInt is changed.
-			key = append(key, codec.IntHandleFlag)
-			key = codec.EncodeInt(key, h.IntValue())
+		// For PartitionHandle on global indexes, we must encode BOTH partition ID and inner handle
+		// in the key to prevent collisions when different partitions have duplicate handles.
+		// This is critical after EXCHANGE PARTITION, which can create duplicate _tidb_rowid values.
```

Comment:
Since the partition ID has been contained in the key, we can remove the one from value part (see `GenIndexValueForClusteredIndexVersion1`). But we have to handle the compatibility issue carefully..

Reply:
I think I will only add it to the key for non-clustered, non-unique global indexes in this PR, and not remove it from the value part, so it is still backward compatible, while allowing duplicate _tidb_rowid's between partitions.

Not sure yet how to plan for removal of the duplicate partition id in the value part. That would also affect the reading logic and would require something like a major version upgrade, or some kind of versioning of the table or global index, to know if the value part can have the partition id or not.